keyboards/gboards/k/buzzsaw/matrix.c

```diff
@@ -0,0 +1,300 @@
+static inline bool store_raw_matrix_row(uint8_t index) {
+  matrix_row_t temp = read_cols(index);
+  if (raw_matrix[index] != temp) {
+    raw_matrix[index] = temp;
+    return true;
+  }
+  return false;
+}
+
+uint8_t matrix_scan(void) {
+    // Then the keyboard
+    if (mcp23018_status) {  // if there was an error
+        if (++mcp23018_reset_loop == 0) {
+            // if (++mcp23018_reset_loop >= 1300) {
+            // since mcp23018_reset_loop is 8 bit - we'll try to reset once in 255 matrix scans
+            // this will be approx bit more frequent than once per second
+            print("trying to reset mcp23018\n");
+            mcp23018_status = init_mcp23018();
+            if (mcp23018_status) {
+                print("left side not responding\n");
+            } else {
+                print("left side attached\n");
+            }
+        }
+    }
+
+    bool changed = false;
+    for (uint8_t i = 0; i < MATRIX_RIGHT; i++) {
+        // select rows from left and right hands
+        uint8_t left_index = i;
+        uint8_t right_index = i + MATRIX_LEFT;
+
+				if  (i < MATRIX_LEFT) select_row(left_index);
+        select_row(right_index);
+
+        // we don't need a 30us delay anymore, because selecting a
+        // left-hand row requires more than 30us for i2c.
+
+        if (i < MATRIX_LEFT) changed |= store_raw_matrix_row(left_index);
+        changed |= store_raw_matrix_row(right_index);
+
+        unselect_rows();
+    }
+
+    debounce(raw_matrix, matrix, MATRIX_ROWS, changed);
+    matrix_scan_quantum();
+
+#ifdef DEBUG_MATRIX
+    for (uint8_t c = 0; c < MATRIX_COLS; c++)
+		for (uint8_t r = 0; r < MATRIX_ROWS; r++)
+		  if (matrix_is_on(r, c)) xprintf("r:%d c:%d \n", r, c);
+#endif
+
+    return 1;
+}
+inline bool matrix_is_on(uint8_t row, uint8_t col) { return (matrix[row] & ((matrix_row_t)1 << col)); }
+inline matrix_row_t matrix_get_row(uint8_t row) { return matrix[row]; }
+void matrix_print(void) {
+    print("\nr/c 0123456789ABCDEF\n");
+    for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
+        phex(row); print(": ");
+        pbin_reverse16(matrix_get_row(row));
+        print("\n");
+    }
+}
+uint8_t matrix_key_count(void) {
+    uint8_t count = 0;
+    for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+        count += bitpop16(matrix[i]);
+    }
+    return count;
+}
+
+// Remember this means ROWS
+static void init_cols(void) {
+    // init on mcp23018
+    // not needed, already done as part of init_mcp23018()
+
+    // Input with pull-up(DDR:0, PORT:1)
+    DDRF  &= ~FMASK;
+    PORTF |=  FMASK;
+}
+static matrix_row_t	read_cols(uint8_t row) {
+    if (row < 7) {
+        if (mcp23018_status) { // if there was an error
+            return 0;
+        } else {
+            uint8_t data = 0;
+            mcp23018_status = i2c_start(I2C_ADDR_WRITE, I2C_TIMEOUT);    if (mcp23018_status) goto out;
+            mcp23018_status = i2c_write(GPIOB, I2C_TIMEOUT);             if (mcp23018_status) goto out;
```

Comment:
```suggestion
      // reading GPIOB (column port) since in mcp23018's sequential mode
      if (mcp23018_status) goto out;	      // it is addressed directly after writing to GPIOA in select_row()
```